# ãƒ—ãƒ­ã‚°ãƒ©ãƒ ãƒ»ãƒãƒ–ï¼ˆProgram Hubï¼›ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã®ä¸­å¿ƒåœ°ï¼‰
extends Node


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# æº–å‚™
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


func _ready():
	self.background_image_cache = MonkeyHand.create(
			self.get_background_artist())	# æ¢ã™å ´æ‰€


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# èƒŒæ™¯ãƒãƒ¼ãƒ‰ã‚’ã¤ã‹ã¿ã«ã„ãçŒ¿ã®æ‰‹
var background_image_cache = null

# å…¨BGM
var cache_dictionary_for_bgm = {}

# å…¨ã‚¤ãƒ©ã‚¹ãƒˆ
var cache_dictionary_for_illust = {}

# å…¨SE
var cache_dictionary_for_se = {}

# å…¨ãƒ†ãƒ­ãƒƒãƒ—
var cache_dictionary_for_telop = {}


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# å¤–ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ç›£ç£å–å¾—
func get_director():
	return $"../../../ğŸŒDirector"


# èƒŒæ™¯ã‚¢ãƒ¼ãƒ†ã‚£ã‚¹ãƒˆ
func get_background_artist():
	return self.get_director().get_node("ğŸŒBackgroundArtist")


# ã‚¤ãƒ©ã‚¹ãƒˆãƒ¬ãƒ¼ã‚¿ãƒ¼å–å¾—
func get_illustrator():
	return self.get_director().get_node("ğŸŒIllustrator")


# BGMå–å¾—
func get_musician_bgm():
	return self.get_director().get_node("ğŸŒMusician_BGM")


# SEå–å¾—
func get_musician_se():
	return self.get_director().get_node("ğŸŒMusician_SE")


# ãƒ†ãƒ­ãƒƒãƒ—ãƒ»ã‚³ãƒ¼ãƒ‡ã‚£ãƒãƒ¼ã‚¿ãƒ¼å–å¾—
func get_telop_coordinator():
	return self.get_director().get_node("ğŸŒTelopCoordinator")


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# å­ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# BGMå–å¾—
func get_bgm(
		target_name):
	return MonkeyHelper.find_node_in_folder(
			target_name,
			func():
				return self.get_musician_bgm(),			# æ¢ã™å ´æ‰€
			func():
				return self.cache_dictionary_for_bgm)	# çµæœã‚’æ ¼ç´ã™ã‚‹å¤‰æ•°


# ã‚¤ãƒ©ã‚¹ãƒˆå–å¾—
func get_illust(
		target_name):	# StringName
	return MonkeyHelper.find_node_in_folder(
			target_name,
			func():
				return self.get_illustrator(),	# æ¢ã™å ´æ‰€
			func():
				return self.cache_dictionary_for_illust)	# çµæœã‚’æ ¼ç´ã™ã‚‹å¤‰æ•°


# åŠ¹æœéŸ³å–å¾—
func get_se(
		target_name):	# StringName
	return MonkeyHelper.find_node_in_folder(
			target_name,
			func():
				return self.get_musician_se(),			# æ¢ã™å ´æ‰€
			func():
				return self.cache_dictionary_for_se)	# çµæœã‚’æ ¼ç´ã™ã‚‹å¤‰æ•°


# ãƒ†ãƒ­ãƒƒãƒ—å–å¾—
func get_telop(
		target_name):	# StringName
	return MonkeyHelper.find_node_in_folder(
			target_name,
			func():
				return self.get_telop_coordinator(),	# æ¢ã™å ´æ‰€
			func():
				return self.cache_dictionary_for_telop)	# çµæœã‚’æ ¼ç´ã™ã‚‹å¤‰æ•°
